### PR TITLE
schedule: introduce ll_schedule_domain

### PR DESCRIPTION
**Pull request by tlauda**
This patch introduces new structure called ll_schedule_domain.
It will be used to initialize ll scheduler with different
types of scheduling time domains. Right now it's only working
based on timer interrupts, but will be extended.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>

### CODE REVIEW

**Comment by tlauda** [comment]
SOFCI TEST

File: src/include/sof/schedule/ll_schedule_source.h [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	trace_event(TRACE_CLASS_LL_SCHED_SRC, format, ##__VA_ARGS__)
+
+#define trace_sch_src_error(format, ...) \
+	trace_error(TRACE_CLASS_LL_SCHED_SRC, format, ##__VA_ARGS__)
+
+#define tracev_sch_src(format, ...) \
+	tracev_event(TRACE_CLASS_LL_SCHED_SRC, format, ##__VA_ARGS__)
+
+struct ll_schedule_source_ops {
+	int (*source_register)(struct ll_schedule_source *src, uint64_t period,
```
Comment (lgirdwood) [inline src/include/sof/schedule/ll_schedule_source.h]: So "source" is the scheduling time domain ? I would rename to "domain" as source and SRC overlap with buffers and SRCs.

File: src/include/user/trace.h [hunk 2]
```diff
@@ -27,39 +27,40 @@ struct system_time {
 } __attribute__((packed));
 
 /* trace event classes - high 8 bits*/
-#define TRACE_CLASS_IRQ		(1 << 24)
```
Comment (lgirdwood) [inline src/include/user/trace.h]: Should be in another commit ?

File: src/arch/xtensa/include/arch/drivers/timer.h [hunk 3]
```diff
@@ -15,18 +15,13 @@
...
 
-struct timer_irq {
-	int logical_irq;
-	void *irq_arg;
-};
-
 struct timer {
 	uint32_t id;
 	int irq;
+	int logical_irq;
```
Comment (lgirdwood) [inline src/arch/xtensa/include/arch/drivers/timer.h]: Can you comment this.

File: src/schedule/timer_source.c [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+#include <stddef.h>
+#include <stdint.h>
+
+struct timer_source {
+	struct timer *timer;
+	uint64_t timeout;
+	void *arg[PLATFORM_CORE_COUNT];
+};
+
+struct ll_schedule_source_ops timer_source_ops;
```
Comment (lgirdwood) [inline src/schedule/timer_source.c]: Is this needed, cant see it used below.

File: src/schedule/timer_source.c [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+#include <sof/lib/alloc.h>
+#include <sof/lib/cpu.h>
+#include <sof/platform.h>
+#include <sof/schedule/ll_schedule_source.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+#include <ipc/topology.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
```
Comment (lgirdwood) [inline src/schedule/timer_source.c]: Is topology needed here ?

File: src/schedule/timer_source.c (hunk 4 above)
Comment (tlauda) [reply src/schedule/timer_source.c]: Used down below in timer_domain_init.

File: src/schedule/timer_source.c (hunk 5 above)
Comment (tlauda) [reply src/schedule/timer_source.c]: Neeeded for `SOF_MEM_CAPS_RAM`.

File: src/include/sof/schedule/ll_schedule_source.h (hunk 1 above)
Comment (tlauda) [reply src/include/sof/schedule/ll_schedule_source.h]: Good suggestion.
Done.

File: src/arch/xtensa/include/arch/drivers/timer.h (hunk 3 above)
Comment (tlauda) [reply src/arch/xtensa/include/arch/drivers/timer.h]: Done.

**Comment by tlauda** [comment]
@lgirdwood Changes applied.

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by tlauda** [comment]
@lgirdwood Waiting for your approval...

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
Good to merge, there is a minor CI checkpatch alignment which can be fixed incrementally. The other CI reports seem to be a false positive (as I can no longer see any errors).

**Comment by xiulipan** [comment]
@lgirdwood Could have a look at the latest QEMU with latest FW, from the log it seems we are trying to read out some timer that is always 0. It seems QEMU did not set them right.

I can see IPC message in the mailbox, but I could not see the IPC is sent.
